sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusListener.cs

```diff
@@ -185,89 +159,89 @@ public void Dispose()
+                if (_batchReceiver != null && _batchReceiver.IsValueCreated)
                 {
-                    Receiver.CloseAsync().Wait();
-                    _receiver = null;
+                    BatchReceiver.CloseAsync().Wait();
+                    _batchReceiver = null;
                 }
 
                 if (_sessionClient != null && _sessionClient.IsValueCreated)
                 {
-                    _sessionClient.Value.CloseAsync().Wait();
+#pragma warning disable AZC0107 // DO NOT call public asynchronous method in synchronous scope.
+                    _sessionClient.Value.DisposeAsync().EnsureCompleted();
+#pragma warning restore AZC0107 // DO NOT call public asynchronous method in synchronous scope.
                     _sessionClient = null;
                 }
 
-                if (_clientEntity != null)
-                {
-                    _clientEntity.CloseAsync().Wait();
-                    _clientEntity = null;
-                }
-
                 _stopAsyncSemaphore.Dispose();
                 _cancellationTokenSource.Dispose();
 
                 _disposed = true;
             }
         }
 
-        private Lazy<MessageReceiver> CreateMessageReceiver()
+        private Lazy<ServiceBusReceiver> CreateMessageReceiver()
         {
-            return new Lazy<MessageReceiver>(() => _messagingProvider.CreateMessageReceiver(_entityPath, _serviceBusAccount.ConnectionString));
+            return new Lazy<ServiceBusReceiver>(() => _messagingProvider.CreateBatchMessageReceiver(_entityPath, _serviceBusAccount.ConnectionString));
         }
 
-        private Lazy<SessionClient> CreateSessionClient()
+        private Lazy<ServiceBusClient> CreateSessionClient()
         {
-            return new Lazy<SessionClient>(() => _messagingProvider.CreateSessionClient(_entityPath, _serviceBusAccount.ConnectionString));
+            return new Lazy<ServiceBusClient>(() => _messagingProvider.CreateSessionClient(_serviceBusAccount.ConnectionString));
         }
 
-        internal async Task ProcessMessageAsync(Message message, CancellationToken cancellationToken)
+        internal async Task ProcessMessageAsync(ProcessMessageEventArgs args)
         {
-            using (CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, _cancellationTokenSource.Token))
+            using (CancellationTokenSource linkedCts = CancellationTokenSource.CreateLinkedTokenSource(args.CancellationToken, _cancellationTokenSource.Token))
             {
-                if (!await _messageProcessor.BeginProcessingMessageAsync(message, linkedCts.Token).ConfigureAwait(false))
+                //TODO consider using internals visible or exposing the Receiver property instead of reflection
+                ServiceBusReceiver receiver = (ServiceBusReceiver) typeof(ProcessMessageEventArgs).GetField("_receiver", BindingFlags.NonPublic | BindingFlags.Instance).GetValue(args);
```

Comment:
I don't think we should get too focused on the reflection perf here. This needs to be fixed to use a public API.

Reply:
Hmm, I think we want to avoid exposing this field. If we can't use reflection or InternalsVisible, we may need to consider just exposing the ProcessMessageEventArgs in Functions as opposed to the receiver. These args have all the message settlement operations. It gets a bit messy when comparing with batch functions as these don't use the Processor and would not have those args. For those functions you could bind to a receiver.